File: java/styx/erebos/erebos-common/src/main/java/indi/xeno/styx/erebos/app/CommandOptions.java
package indi.xeno.styx.erebos.app;

import org.apache.commons.cli.BasicParser;
import org.apache.commons.cli.Option;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;
import org.apache.commons.cli.Parser;

import static org.apache.commons.cli.Option.UNLIMITED_VALUES;

public class CommandOptions {

  private final Options options;

  public static final char VALUE_SEP = '=';

  CommandOptions() {
    options = new Options();
  }

  public CommandOptions addNoArgOption(String opt, String longOpt, String desc) {
    Option option = new Option(opt, longOpt, false, desc);
    options.addOption(option);
    return this;
  }

  public CommandOptions addSingleArgOption(
      String opt, String longOpt, String desc, String argName, boolean required) {
    Option option = new Option(opt, longOpt, true, desc);
    option.setRequired(required);
    option.setArgName(argName);
    options.addOption(option);
    return this;
  }

  public CommandOptions addListArgOption(
      String opt, String longOpt, String desc, String argName, boolean required) {
    Option option = new Option(opt, longOpt, true, desc);
    option.setRequired(required);
    option.setArgName(argName);
    option.setArgs(UNLIMITED_VALUES);
    options.addOption(option);
    return this;
  }

  public CommandOptions addMapArgOption(
      String opt, String longOpt, String desc, String argName, boolean required) {
    Option option = new Option(opt, longOpt, true, desc);
    option.setRequired(required);
    option.setArgName(argName);
    option.setArgs(UNLIMITED_VALUES);
    option.setValueSeparator(VALUE_SEP);
    options.addOption(option);
    return this;
  }

  CommandArguments parse(String[] args) {
    Parser parser = new BasicParser();
    try {
      return new CommandArguments(parser.parse(options, args));
    } catch (ParseException ex) {
      throw new IllegalArgumentException(ex);
    }
  }
}

File: java/styx/hypnos/hypnos-common/src/main/scala/indi/xeno/styx/hypnos/app/CommandOptions.scala
package indi.xeno.styx.hypnos.app

import org.apache.commons.cli.Option.UNLIMITED_VALUES
import org.apache.commons.cli.{DefaultParser, Options, Option => CliOption}

object CommandOptions {

  val VALUE_SEP: Char = '='

  private def arg(
      opt: String,
      longOpt: String,
      desc: String,
      argName: String,
      required: Boolean
  ): CliOption = {
    val option = new CliOption(opt, longOpt, true, desc)
    option.setRequired(required)
    option.setArgName(argName)
    option
  }
}

class CommandOptions private (private val options: Options) {

  private[app] def this() {
    this(new Options())
  }

  def noArg(opt: String, longOpt: String, desc: String): CommandOptions = {
    val option = new CliOption(opt, longOpt, false, desc)
    options.addOption(option)
    this
  }

  def singleArg(
      opt: String,
      longOpt: String,
      desc: String,
      argName: String,
      required: Boolean
  ): CommandOptions = {
    val option = CommandOptions.arg(opt, longOpt, desc, argName, required)
    options.addOption(option)
    this
  }

  def listArg(
      opt: String,
      longOpt: String,
      desc: String,
      argName: String,
      required: Boolean
  ): CommandOptions = {
    val option = CommandOptions.arg(opt, longOpt, desc, argName, required)
    option.setArgs(UNLIMITED_VALUES)
    options.addOption(option)
    this
  }

  def mapArg(
      opt: String,
      longOpt: String,
      desc: String,
      argName: String,
      required: Boolean
  ): CommandOptions = {
    val option = CommandOptions.arg(opt, longOpt, desc, argName, required)
    option.setArgs(UNLIMITED_VALUES)
    option.setValueSeparator(CommandOptions.VALUE_SEP)
    options.addOption(option)
    this
  }

  private[app] def parse(args: Array[String]): CommandArguments = {
    val parser = new DefaultParser()
    new CommandArguments(parser.parse(options, args))
  }
}
